clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java

```diff
@@ -639,7 +652,7 @@ private int estimatedBytesWritten() {
      * the checking should be based on the capacity of the initialized buffer rather than the write limit in order
      * to accept this single record.
      */
-    public boolean hasRoomFor(long timestamp, byte[] key, byte[] value) {
+    public boolean hasRoomFor(long timestamp, ByteBuffer key, ByteBuffer value) {
```

Comment:
Pre-existing: in estimating whether there is room, we don't apply the estimated compression to the inbound record but assume a compression ratio of 1.0. This is conservative, but may be overly conservative. Do you know why this was done? Since it is not obvious and if you know why it would be worth adding a comment (line 670)